File: Theseus/Generators/GeneratorDepthFirst.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Theseus.Data;

namespace Theseus.Generators
{
	public struct ConfiguratorDepthFirst
	{
		public bool RandomTraverse;
	}

	public static class GeneratorDepthFirst
	{
		private static Maze _maze;
		private static ConfiguratorDepthFirst _configurator;

		private static readonly Random Rnd = new Random((int)DateTime.Now.Ticks);
		private static List<Cell> _visited;

		public static void Generate(Maze maze, ConfiguratorDepthFirst configurator)
		{
			// Initialize
			_configurator = configurator;
			_maze = maze;
			_visited = new List<Cell>(maze.Size.X * maze.Size.Y);

			Visit(maze.Exit);
		}

		private static void Visit(Cell cell)
		{
			Stack<Cell> running = new Stack<Cell>(_maze.Size.X * _maze.Size.Y);

			_visited.Add(cell);

			running.Push(cell);
			while (running.Count > 0)
			{
				Cell current = running.Peek();

				List<Cell> neighbors = current.Neighbors;
				if (_configurator.RandomTraverse)
					neighbors.Shuffle();

				bool finished = true;
				foreach (Cell neighbor in neighbors)
				{
					if (neighbor == null) continue;
					if (_visited.Contains(neighbor)) continue;

					finished = false;

					Cell.CreatePath(current, neighbor);
					running.Push(neighbor);
					_visited.Add(neighbor);
					break;
				}

				if (finished)
					running.Pop();
			}
		}

		private static void Shuffle<T>(this IList<T> list)
		{
			int n = list.Count;
			while (n > 1)
			{
				n--;
				int k = Rnd.Next(n + 1);
				T value = list[k];
				list[k] = list[n];
				list[n] = value;
			}
		}
	}
}

File: Core/Generators/GeneratorDepthFirst.cpp
#include "GeneratorDepthFirst.h"
#include "../Maze.h"
#include "../Cell.h"
#include <stack>
#include <algorithm>
#include <time.h>

namespace Theseus { namespace Core {

void GeneratorDepthFirst::generate(Maze* maze, const ConfiguratorDepthFirst& configurator)
{
	GeneratorDepthFirst generator;
	generator._configurator = configurator;
	generator._maze = maze;

	srand((unsigned int)time(NULL));

	generator.visit(maze->getExit());
}

void GeneratorDepthFirst::visit(Cell* cell)
{
	std::stack<Cell*> running;
	
	_visitedCells.push_back(cell);
	running.push(cell);
	while (running.size() > 0)
	{
		Cell* current = running.top();

		std::vector<Cell*> neighbors;
		neighbors.push_back(_maze->getCellNeighbor(current,North));
		neighbors.push_back(_maze->getCellNeighbor(current,East));
		neighbors.push_back(_maze->getCellNeighbor(current,South));
		neighbors.push_back(_maze->getCellNeighbor(current,West));
		if (_configurator.RandomTraverse)
			shuffle(neighbors);

		bool finished = true;
		for (std::vector<Cell*>::iterator it = neighbors.begin(); it != neighbors.end(); ++it)
		{
			Cell* neighbor = *it;

			if (neighbor == NULL) continue;
			if (std::find(_visitedCells.begin(), _visitedCells.end(), neighbor) != _visitedCells.end()) continue;

			finished = false;

			_maze->openPath(current, neighbor);
			running.push(neighbor);
			_visitedCells.push_back(neighbor);
			break;
		}

		if (finished)
			running.pop();
	}
}

void shuffle(std::vector<Cell*>& cells)
{
	size_t n = cells.size();
	while (n > 1)
	{
		n--;
		size_t k = rand() % (n+1);
		Cell* cell = cells[k];
		cells[k] = cells[n];
		cells[n] = cell;
	}
}

}}